### ISSUE TITLE
4.1.0 Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.

### ISSUE THREAD

**Issue by Havunen** [dadhi/DryIoc#205]
DryIoc 4.1.0 is throwing exception for load test. 

See updated pull request to reproduce the problem:
https://github.com/dadhi/DryIoc/pull/165

Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.

```
DryIoc.dll!DryIoc.Interpreter.InterpretGetScopedViaFactoryDelegateNoDisposalIndex(DryIoc.IResolverContext resolver, FastExpressionCompiler.LightExpression.MethodCallExpression callExpr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec) Line 3284
	at C:\git\DryIoc\src\DryIoc\Container.cs(3284)
DryIoc.dll!DryIoc.Interpreter.TryInterpretMethodCall(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, ref object result) Line 2996
	at C:\git\DryIoc\src\DryIoc\Container.cs(2996)
DryIoc.dll!DryIoc.Interpreter.TryInterpret(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, out object result) Line 2653
	at C:\git\DryIoc\src\DryIoc\Container.cs(2653)
DryIoc.dll!DryIoc.Interpreter.TryInterpret(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, out object result) Line 2658
	at C:\git\DryIoc\src\DryIoc\Container.cs(2658)
DryIoc.dll!DryIoc.Interpreter.TryInterpret(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, out object result) Line 2614
	at C:\git\DryIoc\src\DryIoc\Container.cs(2614)
DryIoc.dll!DryIoc.Interpreter.InterpretGetNameScopedViaFactoryDelegate(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.MethodCallExpression callExpr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec) Line 3447
	at C:\git\DryIoc\src\DryIoc\Container.cs(3447)
DryIoc.dll!DryIoc.Interpreter.TryInterpretMethodCall(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, ref object result) Line 3008
	at C:\git\DryIoc\src\DryIoc\Container.cs(3008)
DryIoc.dll!DryIoc.Interpreter.TryInterpret(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, out object result) Line 2653
	at C:\git\DryIoc\src\DryIoc\Container.cs(2653)
DryIoc.dll!DryIoc.Interpreter.TryInterpret(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, out object result) Line 2658
	at C:\git\DryIoc\src\DryIoc\Container.cs(2658)
DryIoc.dll!DryIoc.Interpreter.TryInterpret(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, out object result) Line 2645
	at C:\git\DryIoc\src\DryIoc\Container.cs(2645)
DryIoc.dll!DryIoc.Interpreter.TryInterpretAndUnwrapContainerException(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, bool useFec, out object result) Line 2510
	at C:\git\DryIoc\src\DryIoc\Container.cs(2510)
DryIoc.dll!DryIoc.Container.ResolveAndCacheKeyed(System.Type serviceType, object serviceKey, DryIoc.IfUnresolved ifUnresolved, object scopeName, System.Type requiredServiceType, DryIoc.Request preResolveParent, object[] args) Line 433
	at C:\git\DryIoc\src\DryIoc\Container.cs(433)
DryIoc.dll!DryIoc.Container.ResolveAndCache(System.Type serviceType, DryIoc.IfUnresolved ifUnresolved) Line 284
	at C:\git\DryIoc\src\DryIoc\Container.cs(284)
DryIoc.dll!DryIoc.Container.DryIoc.IResolver.Resolve(System.Type serviceType, DryIoc.IfUnresolved ifUnresolved) Line 267
	at C:\git\DryIoc\src\DryIoc\Container.cs(267)
DryIoc.dll!DryIoc.Resolver.Resolve(DryIoc.IResolver resolver, System.Type serviceType) Line 6996
	at C:\git\DryIoc\src\DryIoc\Container.cs(6996)
LoadTest.exe!LoadTest.Program.IterateInOrder.AnonymousMethod__0() Line 142
	at C:\git\DryIoc\playground\LoadTest\Program.cs(142)
[External Code]
```

**Comment by dadhi** [comment]
Hi @Havunen 

You are just in time, thanks for finding.
Do you have maybe a line numbers for the exception stack?

**Comment by Havunen** [comment]
@dadhi 
Sure. I updated the issue. You can also reproduce the issue with changes done to LoadTest: 
https://github.com/dadhi/DryIoc/pull/165

**Comment by Havunen** [comment]
@dadhi 
When debuggin the edited load test I am also seeing following alert to popup.

![image](https://user-images.githubusercontent.com/2021355/71975347-5e40ce80-321c-11ea-8d74-ef881caa5ab3.png)

**Comment by dadhi** [comment]
> When debugging the edited load test I am also seeing following alert to popup.

Thanks, this is valuable - some of my assumtions were surely wrong. Will check.

**Comment by Havunen** [comment]
Awesome, I'm waiting for next release! :)

**Comment by dadhi** [comment]
@Havunen  Did you see the results I've added in comment in LoadTest Program.cs?
Interesting, is it OK to have 15 and 16 min of test time? I have also saw 155 threads in output. Sorry I did not have much time to fully grasp what the test is doing, just concentrated on the fixes.

**Comment by Havunen** [comment]
Yes I noticed it. The test takes for a while because it creates lots of parallel load to verify there is no dead-lock / locking issues. I will run the tests now and compare it to previous DryIoc version to see if there is performance change.

**Comment by Havunen** [comment]
@dadhi Did you notice that when running LoadTest using "WithoutFastExpressionCompiler" -option its now crashing to new exception: 

```
Unhandled Exception:
Unhandled Exception:

Unhandled Exception:Unhandled Exception: Unhandled Exception:
Unhandled Exception: Unhandled Exception: Unhandled Exception: Unhandled Exception: Unhandled Exception:Unhandled Exception:
 Unhandled Exception: Unhandled Exception:   Unhandled Exception:
Unhandled Exception:


Unhandled Exception: Unhandled Exception: Unhandled Exception: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at DryIoc.Container.Registry.TryCacheKeyedFactory(Type type, Object key, Object factory)
   at DryIoc.Container.ResolveAndCacheKeyed(Type serviceType, Object serviceKey, IfUnresolved ifUnresolved, Object scopeName, Type requiredServiceType, Request preResolveParent, Object[] args)    at DryIoc.Container.ResolveAndCache(Type serviceType, IfUnresolved ifUnresolved)
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved)
   at LoadTest.Program.<>c__DisplayClass3_0.<IterateInOrder>b__0()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at DryIoc.Container.Registry.TryCacheKeyedFactory(Type type, Object key, Object factory)
   at DryIoc.Container.ResolveAndCacheKeyed(Type serviceType, Object serviceKey, IfUnresolved ifUnresolved, Object scopeName, Type requiredServiceType, Request preResolveParent, Object[] args)    at DryIoc.Container.ResolveAndCache(Type serviceType, IfUnresolved ifUnresolved)
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved)
   at LoadTest.Program.<>c__DisplayClass3_0.<IterateInOrder>b__0()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

**Comment by dadhi** [comment]
I did not test it without FEC, but now will do. thanks again. I have a clue regarding this exception.

**Comment by dadhi** [comment]
@Havunen ,
I have fixed the keyed resolution cache which was not working at all  - that means controllers were not cached and not compiled either.
After the fix it is became clear that compilation slows down the whole thing 100x times - so to emulate the previous results I need to add `WithUseInterpreation()` rule.

The problem with compilation is not only compilation time but multiple threads attempting to compile the same expression wasting the resources and moreover competing over resources slowing everythijg even further (cpu was 100% loaded).

**Comment by Havunen** [comment]
@dadhi 

I think its real-world problem in web applications, multiple threads can start resolving same controller same time? It would be good to get it fixed. 

In the load test I tried to compile and cache everything before the actual parallel work starts. 
See these code lines:
https://github.com/dadhi/DryIoc/blob/master/playground/LoadTest/Program.cs#L108-L110

I thought DryIoc works so that first time it resolves controller it interprets the result and then next time it caches it.  Doesn't it work like that anymore?

**Comment by dadhi** [comment]
The lines you've mentioned should work as expected , I will add stopwatch to track how much time it takes. 
I've added the issue #208 to address compilation.

**Comment by dadhi** [comment]
I have documented the exact pipeline here https://github.com/dadhi/DryIoc/issues/199#issuecomment-573291497

**Comment by dadhi** [comment]
I think, here is the way to go for now https://github.com/dadhi/DryIoc/issues/208#issuecomment-573294503

**Comment by dadhi** [comment]
Feeling dumb, I've fixed another issue with the keyed service and it started to work.

**Comment by Havunen** [comment]
I think ideally it would be nice to run LoadTest using all possible Configurations, but then it would take even longer to finish...

**Comment by dadhi** [comment]
I have run it locally with and without FEC and the times are both faster (~ 7 and 10 minutes). I have added the results to the source and will push it later.
It would be also interesting to check `WithUseInterpretation`.. will see.

**Comment by dadhi** [comment]
Here the results:
```cs
/*
* The results after fixes on my 8750h machine:
*
* - Validation finished       - 00:00:27.39
*
* - ResolveAllControllersOnce of 156 controllers
* -- Default               - 1st time: 0.2001231 seconds, 2nd time (cache entry compilation):    4.8156956  seconds (!!! - FEC is 4 times faster)
* -- WithoutFEC            - 1st time: 0.2052857 seconds, 2nd time (cache entry compilation):    16.4215528 seconds
* -- WithUseInterpretation - 1st time: 0.20758   seconds, 2nd time (cache entry interpretation): 0.1651708  seconds
*
* - Starting compiled + cached tests:
* -- Load Test Finished       - 00:00:37.35; WithoutFEC: 00:00:33.65; WithUseInterpretation: 00:00:58.54; 
* -- Randomized Load Finished - 00:07:42.93; WithoutFEC: 00:07:49.95; WithUseInterpretation: 00:08:07.61; 
*
* - Starting cold run tests
* -- Load Test Finished       - 00:00:41.93; WithoutFEC: 00:00:50.20; WithUseInterpretation: 00:00:59.72; 
* -- Randomized Load Finished - 00:07:53.44; WithoutFEC: 00:08:18.08; WithUseInterpretation: 00:08:20.03; 
*/

```

**Comment by Havunen** [comment]
hmm, in the second section 

> * - Starting compiled + cached tests:
> * -- Load Test Finished       - 00:00:37.35; WithoutFEC: 00:00:33.65; WithUseInterpretation: 00:00:58.54; 

does not that mean FEC compiled code executes slower than FEC (linq) compiled code? Randomized tests we can pretty much ignore because there is random involved, one run might resolve heavy constructor more times than next run.

**Comment by dadhi** [comment]
I am not sure, may be I did something on my machine while testing or just exchanged the results by mistake. I run each test only once.

**Comment by Havunen** [comment]
I tested it on my machine too and linq compiled code execute about 25% faster, no idea why :(   IL code seems very simple... maybe there is a catch somewhere

**Comment by dadhi** [comment]
Thanks for the testing
Obviously there is a catch - I will open a separate issue for this.
